### PR TITLE
[FSDP] Reshard frozen params in backward

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103270
* __->__ #101982

This PR makes a first attempt at improving FSDP's fine-tuning support by adding hooks to reshard frozen parameters in the backward pass.
- Without this, frozen parameters involved in gradient computation are kept as unsharded through the entire backward pass.
- The approach is to register a multi-grad ~~post~~-hook on the _input_ activations to the FSDP module, where the hook performs the resharding after all gradients for the FSDP module must have been computed (meaning that we are safe to reshard).

~~This PR relies on adding a "multi-grad post-hook" that differs from the existing "multi-grad hook" from `register_multi_grad_hook()`. I find that with `register_multi_grad_hook()`, sometimes the unit test counting the number of times `_post_backward_reshard()` is called fails (due to it not being called).~~ This was resolved in https://github.com/pytorch/pytorch/pull/102859.

### CODE REVIEW

**Comment by awgu** [comment]
@rohan-varma The catch-all reshard is still useful in the case that the very first input activation does not require gradient. Then, we have no activation on which we can reshard the root FSDP instance's parameters if they are frozen.

**Comment by pengyanghua** [comment]
@awgu @rohan-varma Is this MR ready for merge or not? What is the blocker?

File: test/distributed/fsdp/test_fsdp_fine_tune.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    """Tests fine-tuning cases where some parameters are frozen."""
+
+    NUM_LINEARS = 6
+
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    def _init_seq_module(self) -> nn.Module:
+        torch.manual_seed(42)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_fine_tune.py:42]: should we have a different seed for each rank so grads on each rank are different?

File: test/distributed/fsdp/test_fsdp_fine_tune.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        def _post_backward_reshard_with_count(*args, **kwargs):
+            nonlocal post_backward_reshard_count
+            post_backward_reshard_count += 1
+            return orig_post_backward_reshard(*args, **kwargs)
+
+        with mock.patch(
+            "torch.distributed.fsdp._runtime_utils._post_backward_reshard",
+            _post_backward_reshard_with_count,
+        ):
+            post_backward_reshard_count = 0
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_fine_tune.py:101]: already defined on 90?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -1061,11 +1072,13 @@ def _finalize_params(
         flat_param = handle.flat_param
         if flat_param.requires_grad:
             if hasattr(flat_param, "_post_backward_hook_state"):
+                post_backward_hook_state_len = len(flat_param._post_backward_hook_state)
                 _p_assert(
-                    len(flat_param._post_backward_hook_state) == 2,
+                    post_backward_hook_state_len == 1
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_runtime_utils.py:1077]: why can it be 1 or 2 now?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 4]
```diff
@@ -1360,6 +1373,39 @@ def _register_post_backward_hooks(
         flat_param._post_backward_hook_state = (acc_grad, hook_handle)  # type: ignore[attr-defined]
 
 
+def _register_post_backward_reshard_only_hooks(
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_runtime_utils.py:1376]: looks like now we have 2 paths where params can be resharded in post backward:
- for flat parameters that don't require grad, do it via a hook on the input activations
- for flat parameters that do require grad, do it with the standard post backward hook

could we unify and do both the reshards with the input activations hook?

**Review by rohan-varma — APPROVED** [review]
awesome work!

File: test/distributed/fsdp/test_fsdp_fine_tune.py (hunk 1 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_fine_tune.py:42]: We should use the same seed for model initialization, and then on L148 I have:
```
torch.manual_seed(self.rank + 1)
```

File: test/distributed/fsdp/test_fsdp_fine_tune.py (hunk 2 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_fine_tune.py:101]: Thanks for the catch!

File: torch/distributed/fsdp/_runtime_utils.py (hunk 3 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:1077]: When the parameters are frozen (`requires_grad=False`), we do not have an `AccumulateGrad` object anymore, so the state looks like:
```
handle.flat_param._post_backward_hook_state = hook_handle
```

Normally, it is both the `hook_handle` and the `acc_grad`.

File: torch/distributed/fsdp/_runtime_utils.py (hunk 4 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:1376]: Resharding on the input activation can be later than our existing post-backward hook, which may regress memory. We should keep both paths and view this newly added path as the fallback for the `requires_grad=False` case.

**Comment by jcyk** [comment]
@awgu , thanks for the great work. but I just come across a problem that might be related.
when I freeze some parameters (by setting `requires_grad = False`)
I encounter the following error:
```
File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py", line 2400, in _che
ck_storage_allocated
    _reshard(state, [handle], [free_unsharded_flat_param])
  File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 389, in _
reshard
    self._free_unsharded_flat_param()
  File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py", line 1629, in _fre
e_unsharded_flat_param
    handle.reshard(free_unsharded_flat_param)
  File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py", line 1600, in resh
ard
    _reshard(state, [handle], [free_unsharded_flat_param])
  File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/fsdp/_runtime_utils.py", line 389, in _
reshard
    _p_assert(storage_size > 0, "Expects storage to be allocated")
  File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/utils.py", line 147, in _p_assert
    self._free_unsharded_flat_param()
  File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py", line 1629, in _fre
e_unsharded_flat_param
    handle.reshard(free_unsharded_flat_param)
  File "/usr/local/python/lib/python3.8/site-packages/torch/distributed/fsdp/flat_param.py", line 1600, in resh
ard
    raise AssertionError(s)
AssertionError: Expects storage to be allocated
```


any idea?

I am using the latest nightly version that I could find to support cuda117. That is `torch-2.1.0.dev20230621+cu117-cp38-cp38-linux_x86_64.whl`

**Comment by awgu** [comment]
@jcyk Would it be possible to share a repro?

**Pull request closed without merging**